diff --git a/gst/audioparsers/gstflacparse.c b/gst/audioparsers/gstflacparse.c
index a53b7eb..485d1fe 100644
--- a/gst/audioparsers/gstflacparse.c
+++ b/gst/audioparsers/gstflacparse.c
@@ -1251,6 +1251,7 @@ gst_flac_parse_handle_headers (GstFlacParse * flacparse)
   caps = gst_caps_new_simple ("audio/x-flac",
       "channels", G_TYPE_INT, flacparse->channels,
       "framed", G_TYPE_BOOLEAN, TRUE,
+      "depth", G_TYPE_INT, flacparse->bps,
       "rate", G_TYPE_INT, flacparse->samplerate, NULL);
 
   if (!flacparse->headers)
